flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/FileSplitMonitoringFunction.java

```diff
@@ -0,0 +1,345 @@
+	private Tuple2<Long, List<FileInputSplit>> currentSplitsToFwd;
+
+	private long globalModificationTime;
+
+	private FilePathFilter pathFilter;
+
+	private volatile boolean isRunning = true;
+
+	/**
+	 * This is the {@link Configuration} to be used to initialize the input format at the reader
+	 * (see {@link #open(Configuration)}). In the codebase, whenever {@link #open(Configuration)} is called,
+	 * it is passed a new configuration, thus ignoring potential user-specified parameters. Now, we pass a
+	 * configuration object at the constructor, which is shipped to the remote tasks.
+	 * */
+	private Configuration configuration;
+
+	public FileSplitMonitoringFunction(
+		FileInputFormat<OUT> format, String path, Configuration configuration,
+		WatchType watchType, int readerParallelism, long interval) {
+
+		this(format, path, configuration, FilePathFilter.DefaultFilter.getInstance(), watchType, readerParallelism, interval);
+	}
+
+	public FileSplitMonitoringFunction(
+		FileInputFormat<OUT> format, String path, Configuration configuration,
+		FilePathFilter filter, WatchType watchType, int readerParallelism, long interval) {
+
+		this.format = Preconditions.checkNotNull(format);
+		this.path = Preconditions.checkNotNull(path);
+		this.configuration = Preconditions.checkNotNull(configuration);
+
+		Preconditions.checkArgument(interval >= 100,
+			"The specified monitoring interval is smaller than the minimum allowed one (100 ms).");
+		this.interval = interval;
+
+		this.watchType = watchType;
+
+		this.pathFilter = Preconditions.checkNotNull(filter);
+
+		this.readerParallelism = Math.max(readerParallelism, 1);
+		this.globalModificationTime = Long.MIN_VALUE;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public void open(Configuration parameters) throws Exception {
+		LOG.info("Opening File Monitoring Source.");
+		
+		super.open(parameters);
+		format.configure(this.configuration);
+	}
+
+	@Override
+	public void run(SourceFunction.SourceContext<FileInputSplit> context) throws Exception {
+		FileSystem fileSystem = FileSystem.get(new URI(path));
+
+		switch (watchType) {
+			case REPROCESS_WITH_APPENDED:
+				while (isRunning) {
+					monitorDirAndForwardSplits(fileSystem, context);
+					Thread.sleep(interval);
+				}
+				isRunning = false;
+				break;
+			case PROCESS_ONCE:
+				monitorDirAndForwardSplits(fileSystem, context);
+				isRunning = false;
+				break;
+			default:
+				isRunning = false;
+				throw new RuntimeException("Unknown WatchType" + watchType);
+		}
+	}
+
+	public boolean isRunning() {
+		return this.isRunning;
+	}
+
+	private void monitorDirAndForwardSplits(FileSystem fs, SourceContext<FileInputSplit> context) throws IOException, JobException {
+		final Object lock = context.getCheckpointLock();
+
+		// it may be non-null in the case of a recovery after a failure.
+		if (currentSplitsToFwd != null) {
+			synchronized (lock) {
+				forwardSplits(currentSplitsToFwd, context);
+			}
+		}
+		currentSplitsToFwd = null;
```

Comment:
Ah, maybe scratch that because you are removing from the list and forwarding atomically in `forwardSplits`. Right?